accepted/00065-modularisation.md

```diff
@@ -0,0 +1,90 @@
+- Feature Name: modular_codebase
+- Start Date: 2.1.2020
+- RFC PR: (leave this empty)
+
+# Summary
+[summary]: #summary
+
+This proposal is about improving the code quality, long term maintainablilty and testabilty of Uyuni by
+creating a set of guidelines based on best practices and retroactively applying them to the existing codebase.
+
+
+# Motivation
+[motivation]: #motivation
+
+The codebase of Uyuni has become increasingly hard to understand and maintain. The majority of time is spent on fixing bugs, debugging, fiddling with complex test setups and simply trying to understand what existing code even does (this includes new code). Even the most knowlegable people on the team don't always know how Uyuni would behave in a certain scenario or how its supposed to behave. Those problems are symptoms of the same underlying issue which is only adding new features without any big picture or architecture in mind and the complexity and lack of quality resulting from that. This creates a feedback loop where too much busywork leaves not enough time for proper reseach and development leading to rushed features leading to more bugs, complexity and busywork. This proposal is an attempt at breaking this loop by investing into a maintainable codebase that requires less busywork and allows for more time on the things that matter.
+
+# Detailed design
+[design]: #detailed-design
+
+The core of this proposal will be modulerizing our codebase driven by the following aspects:
```

Comment:
Let's talk about design: do we refer to the module and their interactions as in the mediator pattern? How would the modules talk to each other?

Reply:
No, it is barely at that level yet. For now the idea is to stop creating a spaghetti and switch to lasagna complexity.

Reply:
What would be the interface between modules then? Are the modules going to be talking to each other using a well-defined protocol or using a pattern?
I think it's not a low-level detail but rather a design decision that has to be taken better earlier than later and specified in the RFC.